source/adapters/level_zero/enqueue_native.cpp

```diff
@@ -10,11 +10,75 @@
 
 #include <ur_api.h>
 
+#include "logger/ur_logger.hpp"
 #include "queue.hpp"
+#include "ur_level_zero.hpp"
 
 ur_result_t ur_queue_handle_legacy_t_::enqueueNativeCommandExp(
-    ur_exp_enqueue_native_command_function_t, void *, uint32_t,
-    const ur_mem_handle_t *, const ur_exp_enqueue_native_command_properties_t *,
-    uint32_t, const ur_event_handle_t *, ur_event_handle_t *) {
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    ur_exp_enqueue_native_command_function_t pfnNativeEnqueue, void *data,
+    uint32_t, const ur_mem_handle_t *,
+    const ur_exp_enqueue_native_command_properties_t *,
+    uint32_t NumEventsInWaitList, const ur_event_handle_t *phEventList,
+    ur_event_handle_t *phEvent) {
+  auto Queue = this;
+
+  bool UseCopyEngine = false;
+
+  // Please note that the following code should be run before the
+  // subsequent getAvailableCommandList() call so that there is no
+  // dead-lock from waiting unsubmitted events in an open batch.
+  // The createAndRetainUrZeEventList() has the proper side-effect
+  // of submitting batches with dependent events.
+  //
+  _ur_ze_event_list_t TmpWaitList;
+  UR_CALL(TmpWaitList.createAndRetainUrZeEventList(
+      NumEventsInWaitList, phEventList, Queue, UseCopyEngine));
+
+  // Get a new command list to be used on this call
+  ur_command_list_ptr_t CommandList{};
+  // TODO: Change UseCopyEngine argument to 'true' once L0 backend
+  // support is added
+  UR_CALL(Queue->Context->getAvailableCommandList(
+      Queue, CommandList, UseCopyEngine, NumEventsInWaitList, phEventList));
+
+  // TODO: do we need to create a unique command type for this?
+  ze_event_handle_t ZeEvent = nullptr;
+  ur_event_handle_t InternalEvent;
+  bool IsInternal = phEvent == nullptr;
+  ur_event_handle_t *Event = phEvent ? phEvent : &InternalEvent;
+  UR_CALL(createEventAndAssociateQueue(Queue, Event,
+                                       UR_COMMAND_ENQUEUE_NATIVE_EXP,
+                                       CommandList, IsInternal, false));
+  ZeEvent = (*Event)->ZeEvent;
+  (*Event)->WaitList = TmpWaitList;
+
+  const auto &WaitList = (*Event)->WaitList;
+  const auto &ZeCommandList = CommandList->first;
+  if (WaitList.Length) {
+    ZE2UR_CALL(zeCommandListAppendWaitOnEvents,
+               (ZeCommandList, WaitList.Length, WaitList.ZeEventList));
+  }
+
+  if (!isInOrderQueue()) {
+    queueFinish();
+    // queue finish will execute the command list. Open it again so that
+    // `zeCommandListAppendSignalEvent` can be executed.
+    UR_CALL(Queue->Context->getAvailableCommandList(
+        Queue, CommandList, UseCopyEngine, NumEventsInWaitList, phEventList));
+  }
+  ScopedCommandList Active{Queue, CommandList->first};
+
+  // Execute interop func
+  pfnNativeEnqueue(Queue, data);
+
+  if (!isInOrderQueue()) {
+    queueFinish();
```

Comment:
Once again, this does not make sense to me, what are you "syncing" before executing the command?

Reply:
Unless "Native Enqueue" is closing the command list, then executing the command list, this queue sync will not result in the commands being executed.

Reply:
I've considered that, but queue finish does seem to have a built-in execute for the normal command lists, see [here](https://github.com/oneapi-src/unified-runtime/blob/main/source/adapters/level_zero/queue.cpp#L860).

Reply:
> I've considered that, but queue finish does seem to have a built-in execute for the normal command lists, see [here](https://github.com/oneapi-src/unified-runtime/blob/main/source/adapters/level_zero/queue.cpp#L860).

That executes "open" command lists, ie lists that are setup as "batchable" command lists. If the queue was created with "batching" then they will be batchable. either way, one should always "execute" the list explicitly even for immedicate command lists. The catch in QueueFinish is to handle the pending lists, but it was always intended that one calls "execute" before QueueFinish. The state machine is broken if one calls Append, then Finish without Execute. The call to "execute all open command lists" is more of a curtesy to help applications that failed to execute the commands.

Reply:
Basically, I can guarantee that the error you are seeing is that the commands have not actually been executed because they have not been explicitly called to be executed.

QueueFinish should always be a final call after we call execute, not before, that is incorect and will cause problems with a L0 api set. Ie, you are preforming a costly Sync call each time you call queueFinish, which is a massive performance hit.

You only want to have one call to QueueFinish and only after the execute has been called so that you don't incur the cost of the Cache flush for Queue Sync multiple times.

Reply:
so the 'fix' here is to add:
`UR_CALL(Queue->executeCommandList(CommandList, false));` before every queueFinish we do here, right?

Reply:
Correct, if you need to call QueueFinish, you need to execute the commands in the list first. The best perf will be to call QueueFinish once at the end once the 2 executes are completed.

Reply:
I have changed this to add the call to `UR_CALL(Queue->executeCommandList(CommandList, false));` although I would be surprised if it will make the test pass, since the test is checking if the dependency of UR call -> interop work will give the correct synchronization. I think the problem is with the synchronization *after* the execution of `pfNativeEnqueue()`

Reply:
This is still failing, as I had expected. Any other thoughts what might be going wrong?

Reply:
I'll debug it and propose a fixed version.

Reply:
https://github.com/oneapi-src/unified-runtime/commit/49f0d499a1522ca7286c11bc229bc1cea6623af6

On my PVC system:
```
[==========] Running 4 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 4 tests from urLevelZeroEnqueueNativeCommandTest
[ RUN      ] urLevelZeroEnqueueNativeCommandTest.Success/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Data_Center_GPU_Max_1100__ID0ID_____________________
[       OK ] urLevelZeroEnqueueNativeCommandTest.Success/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Data_Center_GPU_Max_1100__ID0ID_____________________ (27 ms)
[ RUN      ] urLevelZeroEnqueueNativeCommandTest.Dependencies/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Data_Center_GPU_Max_1100__ID0ID_____________________
[       OK ] urLevelZeroEnqueueNativeCommandTest.Dependencies/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Data_Center_GPU_Max_1100__ID0ID_____________________ (372 ms)
[ RUN      ] urLevelZeroEnqueueNativeCommandTest.DependenciesURBefore/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Data_Center_GPU_Max_1100__ID0ID_____________________
[       OK ] urLevelZeroEnqueueNativeCommandTest.DependenciesURBefore/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Data_Center_GPU_Max_1100__ID0ID_____________________ (371 ms)
[ RUN      ] urLevelZeroEnqueueNativeCommandTest.DependenciesURAfter/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Data_Center_GPU_Max_1100__ID0ID_____________________
[       OK ] urLevelZeroEnqueueNativeCommandTest.DependenciesURAfter/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Data_Center_GPU_Max_1100__ID0ID_____________________ (371 ms)
[----------] 4 tests from urLevelZeroEnqueueNativeCommandTest (1144 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test suite ran. (1144 ms total)
[  PASSED  ] 4 tests.
```

Reply:
Thanks @pbalcer I'll apply this commit now

Reply:
It's still failing, I'll debug on the specific system this is failing on the CI.

Reply:
That's great thanks @pbalcer

Reply:
https://github.com/pbalcer/unified-runtime/commit/52cc57d1c9f17b83a75fad0847cd4903b546761f

Arc A750 (CI):
```
[==========] Running 4 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 4 tests from urLevelZeroEnqueueNativeCommandTest
[ RUN      ] urLevelZeroEnqueueNativeCommandTest.Success/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Arc_TM__A750_Graphics__ID0ID____V________________
[       OK ] urLevelZeroEnqueueNativeCommandTest.Success/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Arc_TM__A750_Graphics__ID0ID____V________________ (47 ms)
[ RUN      ] urLevelZeroEnqueueNativeCommandTest.Dependencies/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Arc_TM__A750_Graphics__ID0ID____V________________
[       OK ] urLevelZeroEnqueueNativeCommandTest.Dependencies/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Arc_TM__A750_Graphics__ID0ID____V________________ (752 ms)
[ RUN      ] urLevelZeroEnqueueNativeCommandTest.DependenciesURBefore/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Arc_TM__A750_Graphics__ID0ID____V________________
[       OK ] urLevelZeroEnqueueNativeCommandTest.DependenciesURBefore/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Arc_TM__A750_Graphics__ID0ID____V________________ (736 ms)
[ RUN      ] urLevelZeroEnqueueNativeCommandTest.DependenciesURAfter/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Arc_TM__A750_Graphics__ID0ID____V________________
[       OK ] urLevelZeroEnqueueNativeCommandTest.DependenciesURAfter/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Arc_TM__A750_Graphics__ID0ID____V________________ (741 ms)
[----------] 4 tests from urLevelZeroEnqueueNativeCommandTest (2277 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test suite ran. (2277 ms total)
[  PASSED  ] 4 tests.
```
PVC:
```
[==========] Running 4 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 4 tests from urLevelZeroEnqueueNativeCommandTest
[ RUN      ] urLevelZeroEnqueueNativeCommandTest.Success/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Data_Center_GPU_Max_1100__ID0ID_____________________
[       OK ] urLevelZeroEnqueueNativeCommandTest.Success/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Data_Center_GPU_Max_1100__ID0ID_____________________ (30 ms)
[ RUN      ] urLevelZeroEnqueueNativeCommandTest.Dependencies/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Data_Center_GPU_Max_1100__ID0ID_____________________
[       OK ] urLevelZeroEnqueueNativeCommandTest.Dependencies/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Data_Center_GPU_Max_1100__ID0ID_____________________ (376 ms)
[ RUN      ] urLevelZeroEnqueueNativeCommandTest.DependenciesURBefore/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Data_Center_GPU_Max_1100__ID0ID_____________________
[       OK ] urLevelZeroEnqueueNativeCommandTest.DependenciesURBefore/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Data_Center_GPU_Max_1100__ID0ID_____________________ (374 ms)
[ RUN      ] urLevelZeroEnqueueNativeCommandTest.DependenciesURAfter/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Data_Center_GPU_Max_1100__ID0ID_____________________
[       OK ] urLevelZeroEnqueueNativeCommandTest.DependenciesURAfter/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Data_Center_GPU_Max_1100__ID0ID_____________________ (373 ms)
[----------] 4 tests from urLevelZeroEnqueueNativeCommandTest (1154 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test suite ran. (1155 ms total)
[  PASSED  ] 4 tests.
```

Reply:
Thanks @pbalcer I've applied this latest commit